File: Tests/Legacy Tests/LegacyTestEmitter.m
//
//  LegacyTestEmitter.m
//  Snowplow
//
//  Copyright (c) 2013-2022 Snowplow Analytics Ltd. All rights reserved.
//
//  This program is licensed to you under the Apache License Version 2.0,
//  and you may not use this file except in compliance with the Apache License
//  Version 2.0. You may obtain a copy of the Apache License Version 2.0 at
//  http://www.apache.org/licenses/LICENSE-2.0.
//
//  Unless required by applicable law or agreed to in writing,
//  software distributed under the Apache License Version 2.0 is distributed on
//  an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
//  express or implied. See the Apache License Version 2.0 for the specific
//  language governing permissions and limitations there under.
//
//  Authors: Joshua Beemster
//  Copyright: Copyright (c) 2020 Snowplow Analytics Ltd
//  License: Apache License Version 2.0
//

#import <XCTest/XCTest.h>
#import "SPTrackerConstants.h"
#import "SPEmitter.h"
#import "SPLogger.h"
#import "SPMockEventStore.h"
#import "SPMockNetworkConnection.h"


@interface SPBrokenNetworkConnection : NSObject <SPNetworkConnection>
@end

@implementation SPBrokenNetworkConnection

- (NSArray<SPRequestResult *> *)sendRequests:(NSArray<SPRequest *> *)requests {
    [NSException raise:@"BrokenNetworkConnection" format:@"Fake exception on network connection."];
    return nil;
}

- (NSURL *)url {
    [NSException raise:@"BrokenNetworkConnection" format:@"Fake exception on network connection."];
    return nil;
}

- (SPHttpMethod)httpMethod {
    [NSException raise:@"BrokenNetworkConnection" format:@"Fake exception on network connection."];
    return SPHttpMethodGet;
}

@end

#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wdeprecated-declarations"

@interface LegacyTestEmitter : XCTestCase

@end

@implementation LegacyTestEmitter

NSString *const TEST_SERVER_EMITTER = @"www.notarealurl.com";

- (void)setUp {
    [super setUp];
    [SPLogger setLogLevel:SPLogLevelVerbose];
}

- (void)tearDown {
    [super tearDown];
}

- (void)testEmitterBuilderAndOptions {
    NSString * protocol = @"https";
    
    SPEmitter *emitter = [SPEmitter build:^(id<SPEmitterBuilder> builder) {
        [builder setUrlEndpoint:TEST_SERVER_EMITTER];
        [builder setHttpMethod:SPHttpMethodPost];
        [builder setEmitRange:500];
        [builder setEmitThreadPoolSize:30];
        [builder setByteLimitGet:30000];
        [builder setByteLimitPost:35000];
        [builder setProtocol:SPProtocolHttps];
    }];
    
    NSString * url = [[NSString alloc] initWithFormat:@"%@://%@/com.snowplowanalytics.snowplow/tp2", protocol, TEST_SERVER_EMITTER];
    
    // Test builder setting properly
    
    XCTAssertNil([emitter callback]);
    XCTAssertTrue([[[emitter urlEndpoint] absoluteString] isEqualToString:url]);
    XCTAssertEqual([emitter httpMethod], SPHttpMethodPost);
    XCTAssertEqual([emitter emitRange], 500);
    XCTAssertEqual([emitter emitThreadPoolSize], 30);
    XCTAssertEqual([emitter byteLimitGet], 30000);
    XCTAssertEqual([emitter byteLimitPost], 35000);
    XCTAssertEqual([emitter protocol], SPProtocolHttps);
    
    SPEmitter * customPathEmitter = [SPEmitter build:^(id<SPEmitterBuilder> builder) {
        [builder setUrlEndpoint:TEST_SERVER_EMITTER];
        [builder setHttpMethod:SPHttpMethodPost];
        [builder setCustomPostPath:@"/com.acme.company/tpx"];
        [builder setEmitRange:500];
        [builder setEmitThreadPoolSize:30];
        [builder setByteLimitGet:30000];
        [builder setByteLimitPost:35000];
        [builder setProtocol:SPProtocolHttps];
    }];
    
    NSString * customUrl = [[NSString alloc] initWithFormat:@"%@://%@/com.acme.company/tpx", protocol, TEST_SERVER_EMITTER];
    XCTAssertTrue([[[customPathEmitter urlEndpoint] absoluteString] isEqualToString:customUrl]);
    
    // Test setting variables to new values
    
    [emitter setUrlEndpoint:@"www.test.com"];
    url = [[NSString alloc] initWithFormat:@"%@://www.test.com/com.snowplowanalytics.snowplow/tp2", protocol];
    XCTAssertTrue([[[emitter urlEndpoint] absoluteString] isEqualToString:url]);
    [emitter setHttpMethod:SPHttpMethodGet];
    XCTAssertEqual([emitter httpMethod], SPHttpMethodGet);
    url = [[NSString alloc] initWithFormat:@"%@://www.test.com/i", protocol];
    XCTAssertTrue([[[emitter urlEndpoint] absoluteString] isEqualToString:url]);
    [emitter setEmitRange:1000];
    XCTAssertEqual([emitter emitRange], 1000);
    [emitter setEmitThreadPoolSize:50];
    XCTAssertEqual([emitter emitThreadPoolSize], 50);
    [emitter setByteLimitGet:1000];
    XCTAssertEqual([emitter byteLimitGet], 1000);
    [emitter setByteLimitPost:50];
    XCTAssertEqual([emitter byteLimitPost], 50);
    
    // Test extra functions
    XCTAssertTrue(![emitter getSendingStatus]);
    XCTAssertTrue([emitter getDbCount] >= 0);
    
    // Allow timer to be set
    [[NSRunLoop mainRunLoop] runUntilDate:[NSDate dateWithTimeIntervalSinceNow:1]];
    [emitter resumeTimer];
}

// MARK: - Emitting tests

- (void)testEmitEventWithBrokenNetworkConnectionDoesntFreezeEmitterStatus {
    id<SPNetworkConnection> networkConnection = [SPBrokenNetworkConnection new];
    SPEmitter *emitter = [self emitterWithNetworkConnection:networkConnection bufferOption:SPBufferOptionSingle];
    [emitter addPayloadToBuffer:[self generatePayloads:1].firstObject];
    
    [NSThread sleepForTimeInterval:1];
    
    XCTAssertFalse([emitter getSendingStatus]);
    
    [emitter flush];
}

- (void)testEmitSingleGetEventWithSuccess {
    SPMockNetworkConnection *networkConnection = [[SPMockNetworkConnection alloc] initWithRequestOption:SPHttpMethodGet statusCode:200];
    SPEmitter *emitter = [self emitterWithNetworkConnection:networkConnection bufferOption:SPBufferOptionSingle];
    [emitter addPayloadToBuffer:[self generatePayloads:1].firstObject];
    
    for (int i = 0; i < 10 && ([networkConnection sendingCount] < 1 || [emitter getSendingStatus]); i++) {
        [NSThread sleepForTimeInterval:1];
    }

    XCTAssertEqual(1, networkConnection.previousResults.count);
    XCTAssertEqual(1, networkConnection.previousResults.firstObject.count);
    XCTAssertTrue([networkConnection.previousResults.firstObject.firstObject isSuccessful]);
    XCTAssertEqual(0, [emitter getDbCount]);
    
    [emitter flush];
}

- (void)testEmitSingleGetEventWithNoSuccess {
    SPMockNetworkConnection *networkConnection = [[SPMockNetworkConnection alloc] initWithRequestOption:SPHttpMethodGet statusCode:500];
    SPEmitter *emitter = [self emitterWithNetworkConnection:networkConnection bufferOption:SPBufferOptionSingle];
    [emitter addPayloadToBuffer:[self generatePayloads:1].firstObject];
    
    for (int i = 0; i < 10 && ([networkConnection sendingCount] < 1 || [emitter getSendingStatus]); i++) {
        [NSThread sleepForTimeInterval:1];
    }

    XCTAssertEqual(1, networkConnection.previousResults.count);
    XCTAssertEqual(1, networkConnection.previousResults.firstObject.count);
    XCTAssertFalse([networkConnection.previousResults.firstObject.firstObject isSuccessful]);
    XCTAssertEqual(1, [emitter getDbCount]);
    
    [emitter flush];
}

- (void)testEmitTwoGetEventsWithSuccess {
    SPMockNetworkConnection *networkConnection = [[SPMockNetworkConnection alloc] initWithRequestOption:SPHttpMethodGet statusCode:200];
    SPEmitter *emitter = [self emitterWithNetworkConnection:networkConnection bufferOption:SPBufferOptionSingle];
    
    for (SPPayload *payload in [self generatePayloads:2]) {
        [emitter addPayloadToBuffer:payload];
    }
    
    for (int i = 0; i < 10 && ([networkConnection sendingCount] < 2 || [emitter getSendingStatus]); i++) {
        [NSThread sleepForTimeInterval:1];
    }

    XCTAssertEqual(0, [emitter getDbCount]);
    int totEvents = 0;
    for (NSArray<SPRequestResult *> *results in networkConnection.previousResults) {
        for (SPRequestResult *result in results) {
            XCTAssertTrue(result.isSuccessful);
            totEvents += result.storeIds.count;
        }
    }
    XCTAssertEqual(2, totEvents);
    
    [emitter flush];
}

- (void)testEmitTwoGetEventsWithNoSuccess {
    SPMockNetworkConnection *networkConnection = [[SPMockNetworkConnection alloc] initWithRequestOption:SPHttpMethodGet statusCode:500];
    SPEmitter *emitter = [self emitterWithNetworkConnection:networkConnection bufferOption:SPBufferOptionSingle];

    for (SPPayload *payload in [self generatePayloads:2]) {
        [emitter addPayloadToBuffer:payload];
    }
    
    for (int i = 0; i < 10 && ([networkConnection sendingCount] < 2 || [emitter getSendingStatus]); i++) {
        [NSThread sleepForTimeInterval:1];
    }

    XCTAssertEqual(2, [emitter getDbCount]);
    for (NSArray<SPRequestResult *> *results in networkConnection.previousResults) {
        for (SPRequestResult *result in results) {
            XCTAssertFalse(result.isSuccessful);
        }
    }
    
    [emitter flush];
}

- (void)testEmitSinglePostEventWithSuccess {
    SPMockNetworkConnection *networkConnection = [[SPMockNetworkConnection alloc] initWithRequestOption:SPHttpMethodPost statusCode:200];
    SPEmitter *emitter = [self emitterWithNetworkConnection:networkConnection bufferOption:SPBufferOptionSingle];
    
    [emitter addPayloadToBuffer:[self generatePayloads:1].firstObject];
    
    for (int i = 0; i < 10 && ([networkConnection sendingCount] < 1 || [emitter getSendingStatus]); i++) {
        [NSThread sleepForTimeInterval:1];
    }

    XCTAssertEqual(1, networkConnection.previousResults.count);
    XCTAssertEqual(1, networkConnection.previousResults.firstObject.count);
    XCTAssertTrue([networkConnection.previousResults.firstObject.firstObject isSuccessful]);
    XCTAssertEqual(0, [emitter getDbCount]);
    
    [emitter flush];
}

- (void)testEmitEventsPostAsGroup {
    SPMockNetworkConnection *networkConnection = [[SPMockNetworkConnection alloc] initWithRequestOption:SPHttpMethodPost statusCode:500];
    SPEmitter *emitter = [self emitterWithNetworkConnection:networkConnection bufferOption:SPBufferOptionDefaultGroup];
    
    NSArray<SPPayload *> *payloads = [self generatePayloads:15];
    for (int i = 0; i < 14; i++) {
        [emitter addPayloadToBuffer:payloads[i]];
    }
    
    for (int i = 0; i < 10 && ([networkConnection sendingCount] < 1 || [emitter getSendingStatus]); i++) {
        [NSThread sleepForTimeInterval:1];
    }

    XCTAssertEqual(14, [emitter getDbCount]);
    networkConnection.statusCode = 200;
    NSUInteger prevSendingCount = [networkConnection sendingCount];
    [emitter addPayloadToBuffer:payloads[14]];
    
    for (int i = 0; i < 10 && ([networkConnection sendingCount] - prevSendingCount < 1 || [emitter getSendingStatus]); i++) {
        [NSThread sleepForTimeInterval:1];
    }

    XCTAssertEqual(0, [emitter getDbCount]);
    int totEvents = 0;
    BOOL areGrouped = NO;
    NSArray<NSArray<SPRequestResult *> *> *prevResults =
    [networkConnection.previousResults subarrayWithRange:NSMakeRange(prevSendingCount, networkConnection.previousResults.count - prevSendingCount)];
    for (NSArray<SPRequestResult *> *results in prevResults) {
        for (SPRequestResult *result in results) {
            XCTAssertTrue(result.isSuccessful);
            NSUInteger ids = result.storeIds.count;
            totEvents += ids;
            areGrouped = areGrouped || ids > 1;
        }
    }
    XCTAssertEqual(15, totEvents);
    XCTAssertTrue(areGrouped);
    
    [emitter flush];
}

- (void)testEmitOversizeEventsPostAsGroup {
    SPMockNetworkConnection *networkConnection = [[SPMockNetworkConnection alloc] initWithRequestOption:SPHttpMethodPost statusCode:500];
    SPEmitter *emitter = [self emitterWithNetworkConnection:networkConnection build:^(id<SPEmitterBuilder> builder) {
        [builder setBufferOption:SPBufferOptionDefaultGroup];
        [builder setByteLimitPost:5];
    }];
    
    NSArray<SPPayload *> *payloads = [self generatePayloads:15];
    for (int i = 0; i < 14; i++) {
        [emitter addPayloadToBuffer:payloads[i]];
    }
    
    for (int i = 0; i < 10 && ([networkConnection sendingCount] < 1 || [emitter getSendingStatus]); i++) {
        [NSThread sleepForTimeInterval:1];
    }

    XCTAssertEqual(0, [emitter getDbCount]);
    networkConnection.statusCode = 200;
    NSUInteger prevSendingCount = [networkConnection sendingCount];
    [emitter addPayloadToBuffer:payloads[14]];
    
    for (int i = 0; i < 10 && ([networkConnection sendingCount] - prevSendingCount < 1 || [emitter getSendingStatus]); i++) {
        [NSThread sleepForTimeInterval:1];
    }

    XCTAssertEqual(0, [emitter getDbCount]);
    
    [emitter flush];
}

// MARK: - Emitter builder

- (SPEmitter *)emitterWithNetworkConnection:(id<SPNetworkConnection>)networkConnection bufferOption:(SPBufferOption)bufferOption {
    return [self emitterWithNetworkConnection:networkConnection build:^(id<SPEmitterBuilder> builder) {
        [builder setBufferOption:bufferOption];
    }];
}

- (SPEmitter *)emitterWithNetworkConnection:(id<SPNetworkConnection>)networkConnection build:(void(^)(id<SPEmitterBuilder>builder))buildBlock {
    return [SPEmitter build:^(id<SPEmitterBuilder> builder) {
        [builder setNetworkConnection:networkConnection];
        [builder setBufferOption:SPBufferOptionSingle];
        [builder setEmitRange:200];
        [builder setByteLimitGet:20000];
        [builder setByteLimitPost:25000];
        [builder setEventStore:[SPMockEventStore new]];
        buildBlock(builder);
    }];
}

- (void)testRemovesEventsFromQueueOnNoRetryStatus {
    SPMockNetworkConnection *networkConnection = [[SPMockNetworkConnection alloc] initWithRequestOption:SPHttpMethodGet statusCode:403];
    SPEmitter *emitter = [self emitterWithNetworkConnection:networkConnection bufferOption:SPBufferOptionSingle];

    [emitter addPayloadToBuffer:[self generatePayloads:1].firstObject];
    
    [NSThread sleepForTimeInterval:1];

    XCTAssertEqual(0, [emitter getDbCount]);
    for (NSArray<SPRequestResult *> *results in networkConnection.previousResults) {
        for (SPRequestResult *result in results) {
            XCTAssertFalse(result.isSuccessful);
        }
    }

    [emitter flush];
}

- (void)testFollowCustomRetryRules {
    SPMockNetworkConnection *networkConnection = [[SPMockNetworkConnection alloc] initWithRequestOption:SPHttpMethodGet statusCode:500];
    SPEmitter *emitter = [self emitterWithNetworkConnection:networkConnection bufferOption:SPBufferOptionSingle];
    
    NSMutableDictionary *customRules = [[NSMutableDictionary alloc] init];
    [customRules setObject:@YES forKey:@403];
    [customRules setObject:@NO forKey:@500];
    [emitter setCustomRetryForStatusCodes:customRules];

    [emitter addPayloadToBuffer:[self generatePayloads:1].firstObject];
    
    [NSThread sleepForTimeInterval:1];

    // no events in queue since they were dropped because retrying is disabled for 500
    XCTAssertEqual(0, [emitter getDbCount]);
    
    networkConnection.statusCode = 403;

    [emitter addPayloadToBuffer:[self generatePayloads:1].firstObject];
    
    [NSThread sleepForTimeInterval:1];

    // event still in queue because retrying is enabled for 403
    XCTAssertEqual(1, [emitter getDbCount]);

    [emitter flush];
}

// MARK: - Service methods

- (NSArray<SPPayload *> *)generatePayloads:(int)count {
    NSMutableArray<SPPayload *> *payloads = [NSMutableArray new];
    for (int i = 0; i < count; i++) {
        SPPayload *payload = [SPPayload new];
        [payload addValueToPayload:@(i).description forKey:@"a"];
        [payloads addObject:payload];
    }
    return payloads;
}

@end

#pragma clang diagnostic pop

File: Tests/Legacy Tests/LegacyTestEmitter.swift
//  Copyright (c) 2013-2023 Snowplow Analytics Ltd. All rights reserved.
//
//  This program is licensed to you under the Apache License Version 2.0,
//  and you may not use this file except in compliance with the Apache License
//  Version 2.0. You may obtain a copy of the Apache License Version 2.0 at
//  http://www.apache.org/licenses/LICENSE-2.0.
//
//  Unless required by applicable law or agreed to in writing,
//  software distributed under the Apache License Version 2.0 is distributed on
//  an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
//  express or implied. See the Apache License Version 2.0 for the specific
//  language governing permissions and limitations there under.

import XCTest
@testable import SnowplowTracker

//class BrokenNetworkConnection: NetworkConnection {
//    func sendRequests(_ requests: [Request]) -> [RequestResult] {
//        NSException.raise("BrokenNetworkConnection", format: "Fake exception on network connection.")
//        return nil
//    }
//
//    var urlEndpoint: URL? {
//        NSException.raise("BrokenNetworkConnection", format: "Fake exception on network connection.")
//        return nil
//    }
//
//    var httpMethod: HttpMethodOptions {
//        NSException.raise("BrokenNetworkConnection", format: "Fake exception on network connection.")
//        return .get
//    }
//}

//#pragma clang diagnostic push
//#pragma clang diagnostic ignored "-Wdeprecated-declarations"

let TEST_SERVER_EMITTER = "www.notarealurl.com"

class LegacyTestEmitter: XCTestCase {
    override func setUp() {
        super.setUp()
        Logger.logLevel = .verbose
    }

    override func tearDown() {
        super.tearDown()
    }

    func testEmitterBuilderAndOptions() {
        let `protocol` = "https"

        let emitter = Emitter(
            namespace: "ns1",
            urlEndpoint: TEST_SERVER_EMITTER,
            method: .post,
            protocol: .https
        ) { emitter in
            emitter.byteLimitGet = 30000
            emitter.byteLimitPost = 35000
            emitter.emitRange = 500
            emitter.emitThreadPoolSize = 30
        }

        var url = "\(`protocol`)://\(TEST_SERVER_EMITTER)/com.snowplowanalytics.snowplow/tp2"

        // Test builder setting properly

        XCTAssertNil(emitter.callback)
        XCTAssertTrue((emitter.urlEndpoint == url))
        XCTAssertEqual(emitter.method, .post)
        XCTAssertEqual(emitter.emitRange, 500)
        XCTAssertEqual(emitter.emitThreadPoolSize, 30)
        XCTAssertEqual(emitter.byteLimitGet, 30000)
        XCTAssertEqual(emitter.byteLimitPost, 35000)
        XCTAssertEqual(emitter.protocol, .https)

        let customPathEmitter = Emitter(
            namespace: "ns2",
            urlEndpoint: TEST_SERVER_EMITTER,
            method: .post,
            protocol: .https,
            customPostPath: "/com.acme.company/tpx"
        ) { emitter in
            emitter.byteLimitGet = 30000
            emitter.byteLimitPost = 35000
            emitter.emitRange = 500
        }

        let customUrl = "\(`protocol`)://\(TEST_SERVER_EMITTER)/com.acme.company/tpx"
        XCTAssertTrue((customPathEmitter.urlEndpoint == customUrl))

        // Test setting variables to new values

        emitter.urlEndpoint = "www.test.com"
        url = "\(`protocol`)://www.test.com/com.snowplowanalytics.snowplow/tp2"
        XCTAssertTrue((emitter.urlEndpoint == url))
        emitter.method = .get
        XCTAssertEqual(emitter.method, .get)
        url = "\(`protocol`)://www.test.com/i"
        XCTAssertTrue((emitter.urlEndpoint == url))
        emitter.emitRange = 1000
        XCTAssertEqual(emitter.emitRange, 1000)
        emitter.emitThreadPoolSize = 50
        XCTAssertEqual(emitter.emitThreadPoolSize, 50)
        emitter.byteLimitGet = 1000
        XCTAssertEqual(emitter.byteLimitGet, 1000)
        emitter.byteLimitPost = 50
        XCTAssertEqual(emitter.byteLimitPost, 50)

        // Test extra functions
        XCTAssertFalse(emitter.isSending)
        XCTAssertTrue(dbCount(emitter) >= 0)

        // Allow timer to be set
        RunLoop.main.run(until: Date(timeIntervalSinceNow: 1))
        emitter.resumeTimer()
        
        flush(emitter)
    }

    // MARK: - Emitting tests

//    func testEmitEventWithBrokenNetworkConnectionDoesntFreezeStatus() {
//        let networkConnection = SPBrokenNetworkConnection()
//        let emitter = self.emitter(with: networkConnection, bufferOption: SPBufferOptionSingle)
//        emitter?.addPayload(toBuffer: generatePayloads(1)?.first)
//
//        Thread.sleep(forTimeInterval: 1)
//
//        XCTAssertFalse(emitter?.getSendingStatus())
//
//        emitter?.flush()
//    }

    func testEmitSingleGetEventWithSuccess() {
        let networkConnection = MockNetworkConnection(requestOption: .get, statusCode: 200)
        let emitter = self.emitter(with: networkConnection, bufferOption: .single)
        addPayload(generatePayloads(1).first!, emitter)

        for _ in 0..<10 {
            Thread.sleep(forTimeInterval: 1)
        }

        XCTAssertEqual(1, networkConnection.previousResults.count)
        XCTAssertEqual(1, networkConnection.previousResults.first!.count)
        XCTAssertTrue(networkConnection.previousResults.first!.first!.isSuccessful)
        XCTAssertEqual(0, dbCount(emitter))

        flush(emitter)
    }

    func testEmitSingleGetEventWithNoSuccess() {
        let networkConnection = MockNetworkConnection(requestOption: .get, statusCode: 500)
        let emitter = self.emitter(with: networkConnection, bufferOption: .single)
        addPayload(generatePayloads(1).first!, emitter)

        for _ in 0..<10 {
            Thread.sleep(forTimeInterval: 1)
        }

        XCTAssertEqual(1, networkConnection.previousResults.count)
        XCTAssertEqual(1, networkConnection.previousResults.first!.count)
        XCTAssertFalse(networkConnection.previousResults.first!.first!.isSuccessful)
        XCTAssertEqual(1, dbCount(emitter))

        flush(emitter)
    }

    func testEmitTwoGetEventsWithSuccess() {
        let networkConnection = MockNetworkConnection(requestOption: .get, statusCode: 200)
        let emitter = self.emitter(with: networkConnection, bufferOption: .single)

        for payload in generatePayloads(2) {
            addPayload(payload, emitter)
        }

        for _ in 0..<10 {
            Thread.sleep(forTimeInterval: 1)
        }

        XCTAssertEqual(0, dbCount(emitter))
        var totEvents = 0
        for results in networkConnection.previousResults {
            for result in results {
                XCTAssertTrue(result.isSuccessful)
                totEvents += result.storeIds?.count ?? 0
            }
        }
        XCTAssertEqual(2, totEvents)

        flush(emitter)
    }

    func testEmitTwoGetEventsWithNoSuccess() {
        let networkConnection = MockNetworkConnection(requestOption: .get, statusCode: 500)
        let emitter = self.emitter(with: networkConnection, bufferOption: .single)

        for payload in generatePayloads(2) {
            addPayload(payload, emitter)
        }

        for _ in 0..<10 {
            Thread.sleep(forTimeInterval: 1)
        }

        XCTAssertEqual(2, dbCount(emitter))
        for results in networkConnection.previousResults {
            for result in results {
                XCTAssertFalse(result.isSuccessful)
            }
        }

        flush(emitter)
    }

    func testEmitSinglePostEventWithSuccess() {
        let networkConnection = MockNetworkConnection(requestOption: .post, statusCode: 200)
        let emitter = self.emitter(with: networkConnection, bufferOption: .single)

        addPayload(generatePayloads(1).first!, emitter)

        for _ in 0..<10 {
            Thread.sleep(forTimeInterval: 1)
        }

        XCTAssertEqual(1, networkConnection.previousResults.count)
        XCTAssertEqual(1, networkConnection.previousResults.first?.count)
        XCTAssertTrue(networkConnection.previousResults.first!.first!.isSuccessful)
        XCTAssertEqual(0, dbCount(emitter))

        flush(emitter)
    }

    func testEmitEventsPostAsGroup() {
        let payloads = generatePayloads(15)
        
        let networkConnection = MockNetworkConnection(requestOption: .post, statusCode: 500)
        let emitter = self.emitter(with: networkConnection, bufferOption: .defaultGroup)
        
        for i in 0..<14 {
            addPayload(payloads[i], emitter)
        }

        for _ in 0..<10 {
            Thread.sleep(forTimeInterval: 1)
        }

        XCTAssertEqual(14, dbCount(emitter))
        networkConnection.statusCode = 200
        let prevSendingCount = networkConnection.sendingCount
        addPayload(payloads[14], emitter)

        for _ in 0..<10 {
            Thread.sleep(forTimeInterval: 1)
        }

        XCTAssertEqual(0, dbCount(emitter))
        var totEvents = 0
        var areGrouped = false
        let prevResults = networkConnection.previousResults[prevSendingCount..<networkConnection.previousResults.count]
        for results in prevResults {
            for result in results {
                XCTAssertTrue(result.isSuccessful)
                let ids = result.storeIds?.count ?? 0
                totEvents += ids
                areGrouped = areGrouped || ids > 1
            }
        }
        XCTAssertEqual(15, totEvents)
        XCTAssertTrue(areGrouped)

        flush(emitter)
    }

    func testEmitOversizeEventsPostAsGroup() {
        let networkConnection = MockNetworkConnection(requestOption: .post, statusCode: 500)
        let emitter = self.emitter(with: networkConnection, bufferOption: .defaultGroup)
        emitter.byteLimitPost = 5

        let payloads = generatePayloads(15)
        for i in 0..<14 {
            addPayload(payloads[i], emitter)
        }

        for _ in 0..<10 {
            Thread.sleep(forTimeInterval: 1)
        }

        XCTAssertEqual(0, dbCount(emitter))
        networkConnection.statusCode = 200
        _ = networkConnection.sendingCount
        addPayload(payloads[14], emitter)

        for _ in 0..<10 {
            Thread.sleep(forTimeInterval: 1)
        }

        XCTAssertEqual(0, dbCount(emitter))

        flush(emitter)
    }

    func testRemovesEventsFromQueueOnNoRetryStatus() {
        let networkConnection = MockNetworkConnection(requestOption: .get, statusCode: 403)
        let emitter = self.emitter(with: networkConnection, bufferOption: .single)

        addPayload(generatePayloads(1).first!, emitter)

        Thread.sleep(forTimeInterval: 1)

        XCTAssertEqual(0, dbCount(emitter))
        for results in networkConnection.previousResults {
            for result in results {
                XCTAssertFalse(result.isSuccessful)
            }
        }

        flush(emitter)
    }

    func testFollowCustomRetryRules() {
        let networkConnection = MockNetworkConnection(requestOption: .get, statusCode: 500)
        let emitter = self.emitter(with: networkConnection, bufferOption: .single)

        var customRules: [Int : Bool] = [:]
        customRules[403] = true
        customRules[500] = false
        emitter.customRetryForStatusCodes = customRules

        addPayload(generatePayloads(1).first!, emitter)

        Thread.sleep(forTimeInterval: 1)

        // no events in queue since they were dropped because retrying is disabled for 500
        XCTAssertEqual(0, dbCount(emitter))

        networkConnection.statusCode = 403

        addPayload(generatePayloads(1).first!, emitter)

        Thread.sleep(forTimeInterval: 1)

        // event still in queue because retrying is enabled for 403
        XCTAssertEqual(1, dbCount(emitter))

        flush(emitter)
    }
    
    func testDoesNotRetryFailedRequestsIfDisabled() {
        let networkConnection = MockNetworkConnection(requestOption: .get, statusCode: 500)
        let emitter = self.emitter(with: networkConnection, bufferOption: .single)
        emitter.retryFailedRequests = false

        addPayload(generatePayloads(1).first!, emitter)

        Thread.sleep(forTimeInterval: 1)

        // no events in queue since they were dropped because retrying is disabled
        XCTAssertEqual(0, dbCount(emitter))

        emitter.retryFailedRequests = true

        addPayload(generatePayloads(1).first!, emitter)

        Thread.sleep(forTimeInterval: 1)

        // event still in queue because retrying is enabled
        XCTAssertEqual(1, dbCount(emitter))

        flush(emitter)
    }

    // MARK: - Emitter builder

    func emitter(with networkConnection: NetworkConnection, bufferOption: BufferOption = .single) -> Emitter {
        let emitter = Emitter(networkConnection: networkConnection, namespace: "ns1", eventStore: MockEventStore()) { emitter in
            emitter.bufferOption = bufferOption
            emitter.emitRange = 200
            emitter.byteLimitGet = 20000
            emitter.byteLimitPost = 25000
        }
        return emitter
    }

    // MARK: - Service methods

    func generatePayloads(_ count: Int) -> [Payload] {
        var payloads: [Payload] = []
        for i in 0..<count {
            let payload = Payload()
            payload.addValueToPayload(NSNumber(value: i).description, forKey: "a")
            payloads.append(payload)
        }
        return payloads
    }
    
    private func addPayload(_ eventPayload: Payload, _ emitter: Emitter) {
        InternalQueue.sync {
            emitter.addPayload(toBuffer: eventPayload)
        }
    }
    
    private func flush(_ emitter: Emitter) {
        InternalQueue.sync {
            emitter.flush()
        }
    }
    
    private func dbCount(_ emitter: Emitter) -> Int {
        return InternalQueue.sync {
            emitter.dbCount
        }
    }
}
//#pragma clang diagnostic pop
